Using Arrow Flight RPC to replace Arrow file for data passing in duplicated NltkSentimentOperator

JamieLXZWoolf opened the pull request:
## Introduction

Currently, the duplicated `NltkSentimentOperator` still uses file passing to communicate between JVM and Python, even though we've replaced CSV with Arrow. For every batch, a new Python process is started and run until the computation for this batch is completed. So every time the operator is executed, there will be multiple Python processes to run.

In this PR, I replace this IPC mechanism with [Arrow Flight RPC](https://arrow.apache.org/blog/2019/10/13/introducing-arrow-flight/), which uses [gRPC](https://grpc.io/) as its RPC framework.

**New Feature**: With the new IPC method, very little time is spent on IPC. The vast majority of time is for sentiment computation (in Python). Previously, changing the batch size will dramatically change the execution time, now it will not and the execution time is dramatically reduced.

## Benchmarks

Below is a comparison of three IPC methods for `NltkSentiment` on my own machine. Because of the OS management issue the time used may be slightly different than the actual time it needs when the machine is fully dedicated to this one job.

Test tuple count:1000 rows.

|  **Size** |  | **Tme for each method /s** |  |  |
| :---: | --- | :---: | --- | --- |
|  **ChunkSize (ArrowBatchSize, only for Arrow)** | **BatchSize** | **CSV** | **Arrow File** | **Arrow Flight** |
|  1 | 1 | 1173.75 | 1325.943 | 37.257 |
|  1 | 10 | 125.848 | 138.638 | 14.549 |
|  10 |  |  | 140.499 | 11.122 |
|  10 | 100 | 17.932 | 21.73 | 8.452 |
|  100 |  |  | 22.029 | 8.091 |
|  10 | 500 | 9.031 | 10.53 | 9.661 |
|  100 |  |  | 10.52 | 8.968 |
|  500 |  |  | 10.21 | 9.251 |
|  10 | 1000 | 7.499 | 8.679 | 9.638 |
|  100 |  |  | 8.479 | 8.965 |
|  500 |  |  | 8.738 | 8.871 |
|  1000 |  |  | 8.267 | 8.896 |

Because this (1000) is still a very small number so maybe this cannot prove the actual performance difference. This is primarily to show that using a Python daemon will significantly reduce the time used to invoke processes and other related overheads.

## Mechanism changes
The mechanism used now is an RPC framework. In JVM (implemented in [`ArrowNltkSentimentOperator.java`](https://github.com/Texera/texera/pull/785/files#diff-5345d2ecb0cf1b16355e8db0ec7d9eda)), the operator object contains a `FlightClient` object. In Python script([`arrow_for_nltk_sentiment.py`](https://github.com/Texera/texera/pull/785/files?file-filters%5B%5D=.py&file-filters%5B%5D=.xml#diff-88aafd286140c0fda89ffe8dd3e7b513)), I defined a `FlightServer` class that inherits `pyarrow.flight.FlightServerBase` and implements some of its methods.

The python process serves as the **server** and is started when the operator is opened, shut down when the operator is closed. It implements basic operations required by **FlightServerBase**, along with a custom action "`compute`" that supports NLTK sentiment analysis.
JVM (the operator) includes a **client** that manages all the operations needed to communicate with the server, checking server status, shutting down the server, passing data (as **Arrow Stream**), calling the server to execute sentiment computation, etc.
The python process is executed once for each operator, regardless of the batch number. It serves as a daemon throughout the lifecycle of the operator.

## Code Changes

- `core/dataflow/pom.xml`: 3 new packages are included. `protobuf-java` 3.5.1 must be included, even though it is included in the dependencies of Arrow Flight (another version). It seems the `protobuf-java` version in Arrow Flight's dependencies will result in run-time errors, which might be a bug in current Arrow Flight (it is not very stable now).
- `ArrowNltkSentimentOperator.java`: 
    1. Some new attributes, including the `flightClient` throughout the operator.
    2. In `open()`: Start the server.
    3. In `close()`: Shut down the server.
    4. In `computeClassLabel()`:  deleted codes related to files.
    5. Replaced file passing methods (`writeArrowFile` and `readArrowFile`) with stream methods (`writeArrowStream` and `readArrowStream`). 
- `arrow_for_nltk_sentiment.py`: implemented `FlightServer` to wrap everything in it.

## Detailed explanation
**APIs used**
Below shows mappings of the RPCs in the client and in the server. The default package here is `org.apache.arrow.flight`, in case of confusion.
|  **RPC in the client** | **Counterpart in the server** | **Notes (some are according to the official blog)** |
| --- | --- | --- |
|  `java.util.Iterator<Result> doAction(Action action, CallOption... options)` | `do_action(self, context, action)` | Each (implementation-specific) action is a string (defined in the script). The client is expected to know available actions. When a specific action is called, the server executes the corresponding action and maybe will return any results, i.e. a generalized function call. |
|  `FlightClient.ClientStreamListener startPut(FlightDescriptor descriptor, VectorSchemaRoot root,FlightClient.PutListener metadataListener, CallOption... options)` to start the put, and then the returned `ClientStreamListener` handles the rest. This is primarily wrapped inside the `ArrowNltkSentimentOperator.writeArrowStream(List<Tuple> values)` method. | `do_put(self, context, descriptor, reader, writer)` | Pass Arrow stream from the client to the server. The data must be associated with a `FlightDescriptor`, which can be either a `path` or a `command`. Here the `path` is not actually a path on the disk, but rather an identifier. |
|  `FlightStream getStream(Ticket ticket, CallOption... options) wrapped inside ArrowNltkSentimentOperator.readArrowStream().` | `do_get(self, context, ticket)` | Before getting the stream, the client must first ask the server for available tickets (to the specified dataset) of the specified FlightDescriptor. (by calling `FlightInfo getInfo(FlightDescriptor descriptor,CallOption... options)`) |
|  `Iterable<FlightInfo> listFlights(Criteria criteria,CallOption... options)` | `list_flights(self, context, criteria)` | Getting a list of available datasets on the server. This method is not used here, but might be useful in the future.|
|  `FlightInfo getInfo(FlightDescriptor descriptor,CallOption... options)` | `get_flight_info(self, context, descriptor)` | Returning an “access plan” for a dataset of interest, possibly requiring consuming multiple data streams. This request can accept custom serialized commands containing, for example, your specific application parameters. |


I defined some custom actions on the Python server.
- `"compute"`: to execute the computation of sentiments.
- `"healthcheck"`: to check the status of the server to see if it is running.
- `"shutdown"`: to shutdown the server.

Here is an explanation of the whole process.
1. (in `open()`) When the `ArrowNltkSentimentOperator` is opened, it executes the python script, which constructs a `FlightServer` object which is then up and running in the specified address. The operator calls `flightClient.doAction(new Action("healthcheck"))` to check the status of the server, and then proceeds if successful (otherwise there will be an exception). 
2. (in`writeArrowStream()`) For every batch, the operator converts list of `Tuple`s into Arrow stream data in almost the exact same way as it would when using Arrow file, except now it sends stream to the server with `flightClient.startPut()` and `streamWriter.putNext()`. The server uses `do_put()` to receive data stream and convert it into a `pyarrow.Table` and store it in the server.
3. (in `computeClassLabel()`) For this batch, the operator then calls `flightClient.doAction(new Action("compute"))` to tell the server to compute sentiments of the specific table that was sent earlier. The server executes computation, and returns back a success message.
4. (in `readArrowStream()`) For this batch, the operator then gets the computed sentiment result by calling `flightClient.getStream(ticket)`. (See the above table about tickets.) The reading and conversion process is the same as what it does when using Arrow file.
5. (in `close()`) When all the batches are finished and the operator closes, it issues a `flightClient.doAction(new Action("shutdown"))` call to shut down the server, and also closes the client.

JamieLXZWoolf commented:
**Another change**
It seems to include Arrow Flight in Maven dependencies, `os-maven-plugin` must also be included in `<extensions>` of the `<build>` of `pom.xml`. This is to parse some os-specific dependencies, otherwise, Travis CI won't pass as the desired artifact name will be wrongly parsed.

YinanDanielZhou commented on core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/arrow/ArrowNltkSentimentOperator.java:
```diff
@@ -59,6 +48,11 @@
             );
 
+    // Flight related
+    private final static Location lc = new Location(URI.create("grpc+tcp://localhost:5005"));
```
Rename `lc` to `location`. In general, try not to use abbreviation for variable naming, it is not worth it.

YinanDanielZhou commented on core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/arrow/ArrowNltkSentimentOperator.java:
```diff
@@ -101,6 +95,17 @@ public void open() throws TexeraException {
+            processBuilder.start();
+            // Connect to server
+            flightClient = FlightClient.builder(rootAllocator, lc).build();
+            System.out.println("Flight Client:\t" + new String(flightClient.doAction(new Action("healthcheck")).next().getBody(), StandardCharsets.UTF_8));
```
comment out all unnecessary `System.out.println` and `print` function calls in the two files. But don't delete them because they can be useful for future debugging purposes.

JamieLXZWoolf commented on core/dataflow/src/main/java/edu/uci/ics/texera/dataflow/arrow/ArrowNltkSentimentOperator.java:
```diff
@@ -59,6 +48,11 @@
             );
 
+    // Flight related
+    private final static Location location = new Location(URI.create("grpc+tcp://localhost:5005"));
```
The location of the server, which is a URI. For now, it is fixed.

JamieLXZWoolf commented:
Some more detailed explanation about newly added methods and APIs used:
Basically this PR is an implementation of [this diagram](https://github.com/Texera/texera/issues/765#issuecomment-657540674)(the desired processing method).

The mechanism used now is an RPC framework. In JVM (implemented in [`ArrowNltkSentimentOperator.java`](https://github.com/Texera/texera/pull/785/files#diff-5345d2ecb0cf1b16355e8db0ec7d9eda)), the operator object contains a `FlightClient` object. In Python script([`arrow_for_nltk_sentiment.py`](https://github.com/Texera/texera/pull/785/files?file-filters%5B%5D=.py&file-filters%5B%5D=.xml#diff-88aafd286140c0fda89ffe8dd3e7b513)), I defined a `FlightServer` class that inherits `pyarrow.flight.FlightServerBase` and implements some of its methods.

Below shows mappings of the RPC stubs in the client and in the server. The default package here is `org.apache.arrow.flight`, in case of confusion.
|  **Subroutine(stub) called in the client** | **Counterpart in the server** | **Notes (some are according to the official blog)** |
| --- | --- | --- |
|  `java.util.Iterator<Result> doAction(Action action, CallOption... options)` | `do_action(self, context, action)` | Each (implementation-specific) action is a string (defined in the script). The client is expected to know available actions. When a specific action is called, the server executes the corresponding action and maybe will return any results, i.e. a generalized function call. |
|  `FlightClient.ClientStreamListener startPut(FlightDescriptor descriptor, VectorSchemaRoot root,FlightClient.PutListener metadataListener, CallOption... options)` to start the put, and then the returned `ClientStreamListener` handles the rest. This is primarily wrapped inside the `ArrowNltkSentimentOperator.writeArrowStream(List<Tuple> values)` method. | `do_put(self, context, descriptor, reader, writer)` | Pass Arrow stream from the client to the server. The data must be associated with a `FlightDescriptor`, which can be either a `path` or a `command`. Here the `path` is not actually a path on the disk, but rather an identifier. |
|  `FlightStream getStream(Ticket ticket, CallOption... options) wrapped inside ArrowNltkSentimentOperator.readArrowStream().` | `do_get(self, context, ticket)` | Before getting the stream, the client must first ask the server for available tickets (to the specified dataset) of the specified FlightDescriptor. (by calling `FlightInfo getInfo(FlightDescriptor descriptor,CallOption... options)`) |
|  `Iterable<FlightInfo> listFlights(Criteria criteria,CallOption... options)` | `list_flights(self, context, criteria)` | Getting a list of available datasets on the server. This method is not used here, but might be useful in the future.|
|  `FlightInfo getInfo(FlightDescriptor descriptor,CallOption... options)` | `get_flight_info(self, context, descriptor)` | Returning an “access plan” for a dataset of interest, possibly requiring consuming multiple data streams. This request can accept custom serialized commands containing, for example, your specific application parameters. |

I defined some custom actions on the Python server.
- `"compute"`: to execute the computation of sentiments.
- `"healthcheck"`: to check the status of the server to see if it is running.
- `"shutdown"`: to shutdown the server.

Here is an explanation of the whole process.
1. (in `open()`) When the `ArrowNltkSentimentOperator` is opened, it executes the python script, which constructs a `FlightServer` object which is then up and running in the specified address. The operator calls `flightClient.doAction(new Action("healthcheck"))` to check the status of the server, and then proceeds if successful (otherwise there will be an exception). 
2. (in`writeArrowStream()`) For every batch, the operator converts list of `Tuple`s into Arrow stream data in almost the exact same way as it would when using Arrow file, except now it sends stream to the server with `flightClient.startPut()` and `streamWriter.putNext()`. The server uses `do_put()` to receive data stream and convert it into a `pyarrow.Table` and store it in the server.
3. (in `computeClassLabel()`) For this batch, the operator then calls `flightClient.doAction(new Action("compute"))` to tell the server to compute sentiments of the specific table that was sent earlier. The server executes computation, and returns back a success message.
4. (in `readArrowStream()`) For this batch, the operator then gets the computed sentiment result by calling `flightClient.getStream(ticket)`. (See the above table about tickets.) The reading and conversion process is the same as what it does when using Arrow file.

JamieLXZWoolf commented:
> @JamieLXZWoolf Thanks for adding the detailed explanation, can you do 2 things:
> 
> 1. move the paragraph you just added to the top of this PR, under the "Mechanism changes" section, to make it more visible
> 2. copy the content you just added to corresponding functions in the Java and Python code. It's also good to have some documentation in the code.
> 
> Other than that this PR is good to be merged.

I've moved the details and copied them into the code documentation. Also, I added a **benchmark** section.
